librad/src/git/local/transport.rs

```diff
@@ -0,0 +1,388 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    io::{self, Cursor, Read, Write},
+    path::PathBuf,
+    process::{Child, ChildStdin, ChildStdout, Command, Stdio},
+    sync::{Arc, Mutex, Once, RwLock},
+    thread,
+};
+
+use git2::transport::{Service, SmartSubtransport, SmartSubtransportStream, Transport};
+use thiserror::Error;
+
+use crate::{
+    git::{
+        ext::{into_git_err, RECEIVE_PACK_HEADER, UPLOAD_PACK_HEADER},
+        local::{self, url::LocalUrl},
+        storage::{self, Storage, WithSigner},
+    },
+    keys::SecretKey,
+    paths::Paths,
+    uri::RadUrn,
+};
+
+lazy_static! {
+    static ref SETTINGS: Arc<RwLock<Option<Settings>>> = Arc::new(RwLock::new(None));
+}
+
+#[derive(Clone)]
+pub struct Settings {
+    pub paths: Paths,
+    pub signer: SecretKey,
+}
+
+pub fn register(settings: Settings) {
+    static INIT: Once = Once::new();
+
+    LocalTransportFactory::new().configure(settings);
+    unsafe {
+        INIT.call_once(move || {
+            git2::transport::register(local::URL_SCHEME, move |remote| {
+                Transport::smart(&remote, true, LocalTransportFactory::new())
+            })
+            .unwrap()
```

Comment:
Why is it safe to `unwrap` here?

Reply:
The only error this can possibly return is `libgit2`'s allocator failing to allocate memory. I would presume that this is a legitimate reason to panic without prior notice.

Reply:
Ooops, right. It cannot fail in other ways because it is invoked under `ONCE`, and registering it twice is the only reason why the underlying native call would fail (apart from allocations).
I did not notice.